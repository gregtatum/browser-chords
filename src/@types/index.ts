import { State } from 'src/store/reducers';

/**
 * Re-exports.
 */
export { AppState, State } from 'src/store/reducers';
export * from './app';
import * as App from './app';

export type APIAction<Type extends string, Args, T> =
  | { type: `${Type}-requested`; generation: number; args: Args }
  | { type: `${Type}-received`; generation: number; args: Args; value: T }
  | {
      type: `${Type}-failed`;
      generation: number;
      args: Args;
      // The old value could be retained.
      value?: T;
      error: unknown;
    };

export type ListFilesCache = Map<string, APICalls.ListFiles>;
export type DownloadFileCache = Map<string, APICalls.DownloadFile>;

export namespace APICalls {
  export type ListFiles = APIAction<
    'list-files',
    { path: string },
    Array<App.DropboxFile>
  >;
  export type DownloadFile = APIAction<
    'download-file',
    { path: string },
    App.DownloadedTextFile
  >;
}

export type Action =
  | { type: 'set-dropbox-access-token'; token: string }
  | { type: 'remove-dropbox-access-token' }
  | { type: 'clear-api-cache' }
  | { type: 'set-api-cache' }
  | { type: 'change-active-file'; value: string }
  | { type: 'modify-active-file'; value: string }
  | { type: 'change-view'; value: App.View }
  | APICalls.ListFiles
  | APICalls.DownloadFile;

/**
 * Selectors always take the root state, and return some part of it.
 */
export type Selector<Returns> = (state: State) => Returns;

/**
 * Provide a mechanism to easily define reducers that are bound to the current
 * set of Actions, and enforce the constraint that the first parameter must be
 * the same as the return value.
 *
 * See src/reducers for practical examples of how this is used.
 */
export type Reducer<S> = (state: S, action: Action) => S;

export type Thunk<Returns = void> = (
  dispatch: Dispatch,
  getState: () => State,
) => Returns;

// type DeThunkFn<T extends Thunk<Returns>, Returns> = (thunk: T) => Returns;

// export type DeThunk<
//   T extends Thunk<Returns>,
//   Returns = ReturnType<ReturnType<Thunk>>
// > = ReturnType<
//   DeThunkFn<T, ReturnType<ReturnType<Returns>>
// >;

/**
 * The rest of these pre-fill Redux with all of the configured Actions and middlewares.
 */
type ThunkDispatch = <Returns>(action: Thunk<Returns>) => Returns;
type PlainDispatch = (action: Action) => Action;
export type GetState = () => State;
export type Dispatch = PlainDispatch & ThunkDispatch;
export type Store = {
  dispatch: Dispatch;
  getState(): State;
  subscribe(listener: () => void): unknown;
  replaceReducer(nextReducer: Reducer<State>): void;
};
