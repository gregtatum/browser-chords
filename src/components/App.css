@keyframes fadein {
  from { opacity: 0; }
  to   { opacity: 1; }
}

@keyframes fadeindelay {
  0% { opacity: 0; }
  30% { opacity: 0; }
  100%   { opacity: 1; }
}

body, .appRoot {
  position: absolute;
  height: 100%;
  width: 100%;
  display: flex;
  margin: 0;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  font-family: sans-serif;
}

code {
  padding: 0 4px;
  background-color: #eee;
  border: 1px solid 1px solid #00000033;
  border-radius: 3px;
}

.appView {
  flex: 1;
  width: 100%;
  overflow: hidden;
  display: flex;
  flex-direction: column;
}

.appViewContents {
  position: relative; /* create a stacking context */
  flex: 1;
  overflow-y: auto;
}

.appViewError {
  width: calc(100% - 20px);
  height: 100%;
  position: absolute;
  margin: 10px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: center;
}
