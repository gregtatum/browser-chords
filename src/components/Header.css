.header {
  border-bottom: 1px solid #ddd;
  padding: 9px;
  display: flex;
}

.headerStart {
  flex: 1;
}

.headerActiveFile > * {
  margin-inline-end: 9px;
  display: inline-block;
}

.headerActiveFile > a {
  color: #000;
}

.headerActiveFile > a:hover {
  text-decoration: none;
}
