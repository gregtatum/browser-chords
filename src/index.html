<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
  <title><%= htmlWebpackPlugin.options.title %></title>
  <link rel="preconnect" href="https://fonts.googleapis.com">
  <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
  <link href="https://fonts.googleapis.com/css2?family=Roboto+Mono:wght@400;700&display=swap" rel="stylesheet">
  <link rel="apple-touch-icon" href="apple-touch-icon.png">
  <meta name="apple-mobile-web-app-title" content="Chords">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <!--
    The source code is available as GPLv3:
    https://github.com/gregtatum/browser-chords/
  -->
</head>
<body>
  <script>
    window.__REACT_DEVTOOLS_GLOBAL_HOOK__ = { isDisabled: true }
  </script>
  <script src="/pdf.min.js"></script>
  <div class="appRoot"></div>
  <div id="menus" class="menus"></div>
</body>
</html>
