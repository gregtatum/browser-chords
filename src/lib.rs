use wasm_bindgen::prelude::*;

#[wasm_bindgen]
pub fn add(a: i32, b: i32) -> i32 {
    return a + b;
}

macro_rules! console_log {
    ($($t:tt)*) => {
        {
            web_sys::console::log_1(&format!($($t)*).into())
        }
    }
}

#[wasm_bindgen]
pub fn greet(name: &str) {
    console_log!("Greetings from wasm: {}!", name);
}
