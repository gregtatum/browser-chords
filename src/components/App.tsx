import * as React from 'react';
import { useSelector } from 'react-redux';
import * as $ from 'src/store/selectors';

import './App.css';

export function App() {
  const isInit = useSelector($.getInit);
  if (!isInit) {
    throw new Error('Expected store to be init.');
  }
  return <h1>React is loaded</h1>;
}
