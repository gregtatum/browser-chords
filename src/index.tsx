import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import { createStore } from './store/create-store';
import * as T from 'src/@types';
import { App } from 'src/components/App';
import { maybeMockGoogleAnalytics } from 'src/utils';

import * as A from 'src/store/actions';
import * as $ from 'src/store/selectors';
import { openDB } from 'src/logic/offline-db';

export * as A from 'src/store/actions';
export * as $ from 'src/store/selectors';
export * as T from 'src/@types';

if (process.env.NODE_ENV !== 'test') {
  init();
}

export async function init(): Promise<void> {
  maybeMockGoogleAnalytics();
  initServiceWorker();

  const store = createStore();

  Object.assign(window as any, {
    store,
    dispatch: store.dispatch,
    getState: store.getState,
    $,
    A,
    expireOauth() {
      const oauth = $.getDropboxOauth(store.getState());
      if (!oauth) {
        return;
      }
      store.dispatch(
        A.setDropboxAccessToken(oauth.accessToken, 0, oauth.refreshToken),
      );
    },
  });

  await store.dispatch(openDB());
  mountReact(store);
}

export function createRootApp(store: T.Store): JSX.Element {
  return (
    <Provider store={store as any}>
      <App key="app" />
    </Provider>
  );
}

function mountReact(store: T.Store): void {
  const mountElement = document.createElement('div');
  mountElement.className = 'appRoot';
  const body = document.body;
  if (!body) {
    throw new Error(
      'Attempting to mount the <App> React component but no document body was found.',
    );
  }
  body.appendChild(mountElement);
  ReactDOM.render(createRootApp(store), mountElement);
}

function initServiceWorker() {
  if (
    process.env.NODE_ENV === 'development' ||
    process.env.NODE_ENV === 'test'
  ) {
    return;
  }
  window.addEventListener('load', async () => {
    try {
      const registration = await navigator.serviceWorker.register(
        '/service-worker.js',
      );
      console.log('Service worker registered: ', registration);
    } catch (error) {
      console.log('Service worker registration failed:', error);
    }
  });
}
