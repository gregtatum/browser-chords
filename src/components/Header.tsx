import * as React from 'react';
import * as Redux from 'react-redux';
import { $, A } from 'src';
import * as Router from 'react-router-dom';
import { isAppSettingScrollTop, UnhandledCaseError } from 'src/utils';

import './Header.css';

export function Header() {
  const view = Redux.useSelector($.getView);
  const path = Redux.useSelector($.getActiveFileDisplayPath);
  const [shouldHideHeader, setShouldHideHeader] = React.useState(false);
  const key = view + path;

  const headerStyle: React.CSSProperties = {};
  if (shouldHideHeader) {
    headerStyle.transform = 'translateY(var(--header-transform-y))';
  }

  React.useEffect(function trackScrolling() {
    const { scrollingElement } = document;
    if (!scrollingElement) {
      return () => {};
    }
    let headerPadding: number;
    {
      const headerPaddingStr =
        getComputedStyle(scrollingElement).getPropertyValue('--header-padding');
      if (!headerPaddingStr) {
        throw new Error('Expected to find a headerPadding style');
      }
      headerPaddingStr.replace('px', '');
      headerPadding = parseInt(headerPaddingStr, 10) * 0.5;
    }

    let _prevScroll = 0;
    let _shouldHideHeader = false;

    const onScroll = () => {
      if (isAppSettingScrollTop()) {
        // The app is setting the scrollTop.
        // Note: This doesn't always trigger on navigation events for some reason.
        setShouldHideHeader(false);
        return;
      }
      const { scrollTop } = scrollingElement;
      const dx = scrollTop - _prevScroll;
      _prevScroll = scrollTop;
      if (scrollTop === 0) {
        _shouldHideHeader = false;
        setShouldHideHeader(false);
      } else if (dx > 0) {
        // Scrolling down;
        if (scrollTop > headerPadding && !_shouldHideHeader) {
          _shouldHideHeader = true;
          setShouldHideHeader(true);
        }
      } else {
        // Scrolling up.
        if (
          // iPad registers scrolling when it drags past the end of the document.
          // Ensure the header doesn't come back when that happens.
          scrollingElement.scrollHeight - scrollTop > window.innerHeight &&
          _shouldHideHeader
        ) {
          _shouldHideHeader = false;
          setShouldHideHeader(false);
        }
      }
    };

    document.addEventListener('scroll', onScroll);
    return () => {
      document.removeEventListener('scroll', onScroll);
    };
  }, []);

  let title = (
    <div className="headerTitle" key={key}>
      <span>🎵</span>
      <span className="headerTitleTitle">Browser Chords</span>
    </div>
  );
  switch (view) {
    case null:
      break;
    case 'settings':
      title = <Path path="/" key={key} title="⚙️ Settings" />;
      break;
    case 'privacy':
      title = <Path path="/" key={key} title="👀 Privacy Policy and Usage" />;
      break;
    case 'view-file':
    case 'view-pdf':
    case 'view-image':
      title = <Path key={key} path={path} />;
      break;
    case 'list-files':
      if (location.pathname !== '/folder' && location.pathname !== '/') {
        title = <Path key={key} path={path} />;
      }
      break;
    default:
      throw new UnhandledCaseError(view, 'View');
  }

  return (
    <div className="header" style={headerStyle}>
      <div className="headerStart">{title}</div>
      <div className="headerEnd">
        <SaveFileButton />
        <RequestFullScreen />
        <SettingsButton />
      </div>
    </div>
  );
}

function goFullScreen() {
  const element = document.querySelector('[data-fullscreen]');
  if (!element) {
    console.error('No [data-fullscreen] was found.');
    return;
  }
  if (element.requestFullscreen) {
    element.requestFullscreen();
  }
  if ((element as any).webkitRequestFullscreen) {
    (element as any).webkitRequestFullscreen();
  }
}

function fullScreenEventHandler(event: KeyboardEvent) {
  if (
    event.key === 'f' &&
    !event.shiftKey &&
    !event.ctrlKey &&
    !event.metaKey &&
    !event.altKey
  ) {
    goFullScreen();
  }
}

function RequestFullScreen() {
  const canGoFullScreen = Redux.useSelector($.canGoFullScreen);
  React.useEffect(() => {
    if (canGoFullScreen) {
      document.addEventListener('keyup', fullScreenEventHandler);
    } else {
      document.removeEventListener('keyup', fullScreenEventHandler);
    }
    return () => {
      document.removeEventListener('keyup', fullScreenEventHandler);
    };
  }, [canGoFullScreen]);

  if (!canGoFullScreen) {
    return null;
  }

  return (
    <button type="button" className="button" onClick={goFullScreen}>
      Fullscreen
    </button>
  );
}

function SettingsButton() {
  const view = Redux.useSelector($.getView);
  if (view === 'settings') {
    return null;
  }
  return (
    <a href="/settings" className="button headerSettings">
      Settings
    </a>
  );
}

function SaveFileButton() {
  const text = Redux.useSelector($.getModifiedText);
  const view = Redux.useSelector($.getView);
  const dispatch = Redux.useDispatch();
  const path = Redux.useSelector($.getPath);
  const request = Redux.useSelector($.getDownloadFileCache).get(path);

  if (!text || view !== 'view-file' || !request) {
    return null;
  }

  return (
    <button
      className="button headerSaveFile"
      onClick={() => {
        dispatch(A.saveFile(path, text, request));
      }}
    >
      Save
    </button>
  );
}

function Path({ path, title }: { path: string; title?: string }) {
  const songTitle = Redux.useSelector($.getActiveFileSongTitleOrNull);

  const breadcrumbs = [];
  let pathGrow = '';
  const parts = path.split('/');
  const fileName = parts.pop();
  for (const part of parts) {
    if (breadcrumbs.length === 0) {
      breadcrumbs.push(
        <Router.Link key="/" to="/">
          Home
        </Router.Link>,
      );
    } else {
      pathGrow += '/' + part;
      breadcrumbs.push(
        <span key={pathGrow + '»'}>»</span>,
        <Router.Link key={pathGrow} to={`/folder${pathGrow}`}>
          {part}
        </Router.Link>,
      );
    }
  }
  const backParts = path.split('/');
  backParts.pop();

  return (
    <>
      <div className="headerPath headerPathFull" key={'full' + path}>
        <span>🎵</span>
        {breadcrumbs}
        <span>»</span>
        {songTitle ?? fileName ? (
          <span title={fileName}>{songTitle ?? fileName}</span>
        ) : null}
        {title ? <span>{title}</span> : null}
      </div>
      <div className="headerPath headerPathMobile" key={'mobile' + path}>
        <Router.Link
          to={`/folder${backParts.join('/')}`}
          className="headerPathBack"
          aria-label="back"
        ></Router.Link>
      </div>
    </>
  );
}
