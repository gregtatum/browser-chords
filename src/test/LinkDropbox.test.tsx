import { createStore } from 'src/store/create-store';
import { App } from 'src/components/App';
import { render, screen, waitFor } from '@testing-library/react';
import * as React from 'react';
import { Provider } from 'react-redux';
import { A } from 'src';

describe('LinkDropbox', () => {
  it('can render', async () => {
    const store = createStore();
    store.dispatch(A.changeFileSystem('dropbox'));
    const { container } = render(
      <Provider store={store as any}>
        <App />
      </Provider>,
    );

    await waitFor(() => screen.getByText(/Store Files on Dropbox/));

    const button = screen.getByText(/Connect Dropbox/);
    await waitFor(() => expect(button.getAttribute('href')).toBeTruthy());

    expect(container.firstChild).toMatchSnapshot();
  });
});
