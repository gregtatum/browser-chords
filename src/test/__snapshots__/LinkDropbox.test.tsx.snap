// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkDropbox can render 1`] = `
<div
  class="linkDropboxHome"
>
  <div
    class="linkDropboxBegin"
  >
    <div
      class="linkDropbox"
    >
      <div
        class="linkDropboxDescription"
      >
        <h1
          class="linkDropboxH1"
        >
          ðŸŽµ
           
          Browser Chords
        </h1>
        <h2
          class="linkDropboxH2"
        >
          An
           
          <a
            href="https://en.wikipedia.org/wiki/IndieWeb"
          >
            Indie Web
          </a>
           
          project by Greg Tatum
        </h2>
        <p>
          Manage tabs, chords, and sheet music in Dropbox. Access and edit them directly in the browser, from anywhere.
        </p>
        <p>
          Privacy is important. This app will only be given access to the 
          <code>
            Dropbox/Apps/Chords
          </code>
           folder in Dropbox to manage files. See the
           
          <a
            href="/privacy"
          >
            privacy policy
          </a>
           for more details. The source code is on
           
          <a
            href="https://github.com/gregtatum/browser-chords"
          >
            GitHub
          </a>
          .
        </p>
      </div>
      <div>
        <a
          class="linkDropboxConnect"
          href="https://www.dropbox.com/oauth2/authorize?response_type=code&code_challenge_method=S256&client_id=FAKE_DROPBOX_CLIENT_ID&code_challenge=fakebase64&redirect_uri=http%3A%2F%2Flocalhost%2Flogin&token_access_type=offline"
        >
          Connect Dropbox
        </a>
      </div>
    </div>
  </div>
  <div
    class="linkDropboxEnd"
  >
    <div
      class="linkDropboxScroller"
    >
      <img
        height="1500"
        src="/Chords.tiny.png"
        width="1125"
      />
      <img
        height="1500"
        src="/Songs.tiny.png"
        width="1125"
      />
      <img
        height="1500"
        src="/Sheet.tiny.png"
        width="1125"
      />
    </div>
  </div>
</div>
`;
